usr/src/common/zfs/zfs_namecheck.c

```diff
@@ -136,7 +136,7 @@ int
 dataset_namecheck(const char *path, namecheck_err_t *why, char *what)
```

Comment:
Since you relaxed the checks in dataset_namecheck(), I think you should modify the dataset_namecheck() callers and add check for '#' there. For example in zfs_ioc_recv().